### PR TITLE
fix(RTC): Set transceiver direction after RTCRtpSender#replaceTrack.

### PR DESCRIPTION
**Pull request by jallamsetty1**
This fixes the issue where TRACK_REMOVED event is not fired when a remote track is removed from the peerconnection.
Fixes https://github.com/jitsi/lib-jitsi-meet/issues/1612 and https://github.com/jitsi/jitsi-meet/issues/8482.

### CODE REVIEW

File: modules/RTC/TPCUtils.js [hunk 1]
```diff
@@ -316,7 +316,7 @@ export class TPCUtils {
     replaceTrack(oldTrack, newTrack) {
         if (oldTrack && newTrack) {
             const mediaType = newTrack.getType();
-            const stream = newTrack.getOriginalStream();
+            const stream = newTrack.stream;
```
Comment (paweldomas) [inline modules/RTC/TPCUtils.js:319]: `.stream` is not really public and should use `getOriginalStream` getter

File: modules/RTC/TPCUtils.js [hunk 2]
```diff
@@ -328,11 +328,11 @@ export class TPCUtils {
...
-            const track = mediaType === MediaType.AUDIO
-                ? stream.getAudioTracks()[0]
-                : stream.getVideoTracks()[0];
-            const transceiver = this.pc.peerconnection.getTransceivers()
-                .find(t => t.receiver.track.kind === mediaType && !t.stopped);
+
+            const transceiver = oldTrack.isVideoTrack() && browser.doesVideoMuteByStreamRemove && oldTrack.isMuted()
+                ? this._findTransceiver(mediaType)
+                : this._findTransceiver(mediaType, oldTrack);
+            const track = mediaType === MediaType.AUDIO ? stream.getAudioTracks()[0] : stream.getVideoTracks()[0];
```
Comment (paweldomas) [inline modules/RTC/TPCUtils.js:335]: why do you need to fetch the `track` this way? shouldn't it be accessible from the `newTrack` via `getTrack()` getter?
Btw. naming of jitsi tracks and webrtc media stream/tracks is very confusing. Maybe it would help to rename the arguments to `oldJitsiTrack` and `newJitsiTrack`.

File: modules/RTC/TPCUtils.js [hunk 3]
```diff
@@ -357,14 +357,35 @@ export class TPCUtils {
...
+
+                    // Remove the old track from the list of local tracks.
                     this.pc.localTracks.delete(oldTrack.rtcId);
                     this.pc.localSSRCs.delete(oldTrack.rtcId);
                 });
         } else if (newTrack && !oldTrack) {
             const ssrc = this.pc.localSSRCs.get(newTrack.rtcId);
 
             return this.addTrackUnmute(newTrack)
+                .then(() => this.setEncodings(newTrack))
```
Comment (paweldomas) [inline modules/RTC/TPCUtils.js:377]: is `setEncodings` needed for the `if (oldTrack && newTrack)` case as well?

File: modules/RTC/TraceablePeerConnection.js [hunk 4]
```diff
@@ -315,14 +315,15 @@ export default function TraceablePeerConnection(
         this.peerconnection.onremovestream
             = event => this._remoteStreamRemoved(event.stream);
     } else {
-        this.peerconnection.ontrack = event => {
-            const stream = event.streams[0];
+        this.onTrack = evt => {
```
Comment (paweldomas) [inline modules/RTC/TraceablePeerConnection.js:318]: why this change?

File: modules/sdp/LocalSdpMunger.js [hunk 5]
```diff
@@ -189,15 +190,22 @@ export default class LocalSdpMunger {
...
             }
         }
+
+        // If the msid attribute is missing, then remove the ssrc from the transformed description so that a
+        // source-remove is signaled to Jicofo. This happens when the direction of the transceiver (or m-line)
+        // is set to 'inactive' or 'recvonly' on Firefox.
+        const msid = mediaSection.ssrcs.find(s => s.attribute === 'msid');
+
+        if (!msid && browser.isFirefox()) {
+            mediaSection.ssrcs = undefined;
```
Comment (paweldomas) [inline modules/sdp/LocalSdpMunger.js:207]: it would be really great to have unit tests for corner cases like this one

File: modules/RTC/TPCUtils.js (hunk 1 above)
Comment (jallamsetty1) [reply modules/RTC/TPCUtils.js:319]: This wasn't an intentional change, was debugging something and forgot to take it out. Thanks for pointing it out.

File: modules/RTC/TPCUtils.js (hunk 3 above)
Comment (jallamsetty1) [reply modules/RTC/TPCUtils.js:377]: It's not needed if there was a source already and only the stream is being replaced on the sender. However, if the client starts muted and the source is being added on unmute, then the encodings need to be configured.

File: modules/RTC/TraceablePeerConnection.js (hunk 4 above)
Comment (jallamsetty1) [reply modules/RTC/TraceablePeerConnection.js:318]: To avoid running into issues when something else like callstats/rtc stats accidentally overrides these methods.

File: modules/sdp/LocalSdpMunger.js [hunk 6]
```diff
@@ -203,9 +203,29 @@ export default class LocalSdpMunger {
...
+            const sources = [ ...new Set(mediaSection.mLine?.ssrcs?.map(s => s.id)) ];
+
+            for (const source of sources) {
+                const msidExists = mediaSection.ssrcs
+                    .find(ssrc => ssrc.id === source && ssrc.attribute === 'msid');
+
+                !msidExists && mediaSection.ssrcs.push({
+                    id: source,
+                    attribute: 'msid',
+                    value: `${this.localEndpointId}-${mediaSection.mLine?.type}-${pcId} ${trackId}-${pcId}`
```
Comment (paweldomas) [inline modules/sdp/LocalSdpMunger.js:226]: can we extract msid generation into a function? it seems to be duplicated with the logic at line 190?

**Review by paweldomas — APPROVED** [review]
